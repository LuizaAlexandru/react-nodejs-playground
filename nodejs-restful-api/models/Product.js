const mongoose = require('mongoose'),
  Schema = mongoose.Schema,
  options = { discriminatorKey: 'type' },

 productSchema = new Schema({
   name: String,
   description: String,
   price: {
    amount: Number,
    currency: String
  },
  images: [],
  package: String,
  departments: [{type: Schema.Types.ObjectId, ref: 'Department'}],
  status: String
}, options);

mongoose.model('Product', productSchema);

module.exports = mongoose.model('Product');
