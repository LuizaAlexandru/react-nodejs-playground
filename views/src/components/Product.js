import React, {Component} from 'react';
import Action from './Action.js';

export default class Product extends Component {
  constructor(props) {
    super(props);

    this.state = props;
  }

  render() {
    const product = this.state.data,
      images = product.images,
      image = product.images && product.images[0];
    return (
      <div className={`product ${this.props.mode}`}>
        <div className="product-header">
          <div className="product-title" title={product.name}>{product.name}</div>
        </div>
        <div className="product-properties">
          <div className="product-price">Price:
          {
          product.price && product.price.amount && product.price.currency ?
            product.price.amount + ' ' + product.price.currency
            : product.price ? product.price + 'EUR':''
          }
          </div>
          {product.size && <div className="product-size">Size: {product.size}</div>}
          {product.power && <div className="product-power">Power: {product.power}</div>}
        </div>
        <div className="product-details">
          <div className="product-image">
            <img className="product-img" src={image && image.src ? image.src : 'http://placehold.it/150x150/637d9f/ffffff'} width={150} height={150} />
          </div>
          <div className="product-actions">
            <Action text={this.props.actionText} job={this.props.actionJob} id={product._id}
              addProduct={this.props.addProduct} removeProduct={this.props.removeProduct} />
          </div>
        </div>
      </div>
    );
  }
}
