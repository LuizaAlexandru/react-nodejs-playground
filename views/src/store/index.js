import {createStore, applyMiddleware} from "redux";
import rootReducer from "../reducers/index";

const asyncDispatchMiddleware = store => next => action => {
  let syncActivityFinished = false;
  let actionQueue = [];

  function flushQueue() {
    actionQueue.forEach(a => store.dispatch(a)); // flush queue
    actionQueue = [];
  }

  function asyncDispatch(asyncAction) {
    actionQueue = actionQueue.concat([asyncAction]);

    if (syncActivityFinished) {
      flushQueue();
    }
  }

  const actionWithAsyncDispatch =
      Object.assign({}, action, { asyncDispatch });

  next(actionWithAsyncDispatch);
  syncActivityFinished = true;
  flushQueue();
};

const store = createStore(
  rootReducer,
  applyMiddleware(asyncDispatchMiddleware)
);

export default store;
