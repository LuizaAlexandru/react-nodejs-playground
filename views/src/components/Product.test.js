import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import Product from './Product.jsx';
import store from '../store/index';

const data = {
  "name":"White T-shirt Elevate",
  "price": {"amount": 17, "currency": "usd"},
  "images":[{"src":"./demo_imgs/shirt.jpg","alt":"test"}],
  "departments":["5b7bbfc0e5b2f20e8cb06866"],
  "_id":"5b6830856882617d10f2a567",
  "type":"Clothing",
  "size":"Small",
  "power": 0
};

test('product toJson matches snapshot', () => {

  const component = renderer.create(
    <Provider store={store}>
      <Product data={data} />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('data is used in tile', () => {
  const component = mount(
  <Provider store={store}>
    <Product data={data} />
  </Provider>);

  expect(component.find('.product-title').text()).toEqual('White T-shirt Elevate');
});
