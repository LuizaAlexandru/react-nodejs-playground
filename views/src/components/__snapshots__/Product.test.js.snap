// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product toJson matches snapshot 1`] = `
<div
  className="product undefined"
>
  <div
    className="product-header"
  >
    <div
      className="product-title"
      title="White T-shirt Elevate"
    >
      White T-shirt Elevate
    </div>
  </div>
  <div
    className="product-properties"
  >
    <div
      className="product-price"
    >
      Price:
      17 usd
    </div>
    <div
      className="product-size"
    >
      Size: 
      Small
    </div>
    0
  </div>
  <div
    className="product-details"
  >
    <div
      className="product-image"
    >
      <img
        className="product-img"
        height={150}
        src="./demo_imgs/shirt.jpg"
        width={150}
      />
    </div>
    <div
      className="product-actions"
    >
      <div
        className="action"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
